client/app/bundles/course/reference-timelines/components/CreateRenameTimelinePrompt.tsx

```diff
@@ -0,0 +1,146 @@
+import { useState } from 'react';
+import { toast } from 'react-toastify';
+import { Alert } from '@mui/material';
+import { TimelineData } from 'types/course/referenceTimelines';
+
+import Prompt from 'lib/components/core/dialogs/Prompt';
+import TextField from 'lib/components/core/fields/TextField';
+import useTranslation from 'lib/hooks/useTranslation';
+
+import { useLastSaved, useSetLastSaved } from '../contexts';
+import { createTimeline, renameTimeline } from '../operations';
+import { useAppDispatch } from '../store';
+import translations from '../translations';
+
+interface CreateRenameTimelinePromptProps {
+  open: boolean;
+  onClose: () => void;
+  renames?: TimelineData;
+  disabled?: boolean;
+}
+
+const CreateRenameTimelinePrompt = (
+  props: CreateRenameTimelinePromptProps,
+): JSX.Element => {
+  const { renames: timeline } = props;
+
+  const { t } = useTranslation();
+
+  const dispatch = useAppDispatch();
+
+  const [newTitle, setNewTitle] = useState(timeline?.title ?? '');
+  const [submitted, setSubmitted] = useState(false);
+
+  const { status } = useLastSaved();
+  const { startLoading, abortLoading, setLastSavedToNow } = useSetLastSaved();
+
+  const isInvalidTitle = submitted && newTitle === '';
+
+  const resetPrompt = (): void => {
+    setNewTitle(timeline?.title ?? '');
+    setSubmitted(false);
+  };
+
+  const handleCreateTimeline = (): void => {
+    startLoading();
+
+    dispatch(createTimeline(newTitle))
+      .then(() => {
+        props.onClose();
+        setLastSavedToNow();
+      })
+      .catch((error) => {
+        abortLoading();
+        toast.error(
+          error ?? t(translations.errorCreatingTimeline, { newTitle }),
+        );
+      });
+  };
+
+  const handleRenameTimeline = (): void => {
+    if (!timeline) throw new Error(`Trying to rename ${timeline} timeline.`);
+
+    if (newTitle !== timeline.title) {
+      startLoading();
+
+      dispatch(renameTimeline(timeline.id, newTitle))
+        .then(() => {
+          props.onClose();
+          setLastSavedToNow();
+        })
+        .catch((error) => {
+          abortLoading();
+
+          toast.error(
+            error ?? t(translations.errorRenamingTimeline, { newTitle }),
+          );
+        });
+    } else {
+      props.onClose?.();
+    }
+  };
+
+  const handleConfirmTitle = (): void => {
+    setSubmitted(true);
+    if (newTitle === '') return;
+
+    if (timeline) {
+      handleRenameTimeline();
+    } else {
+      handleCreateTimeline();
+    }
+  };
+
+  return (
+    <Prompt
```

Comment:
I think this is because there is a split second between when the `startLoading` is invoked and actually completed. It is not easily reproduced, not even with CPU throttling enabled.

I suspect this might be caused by slow re-renders of `AssignableTimeline` and `AssignedTimeline` that subscribe to `status` from `LastSavedContext`. Ultimately, the solution is to virtualise `TimelinesStack` and `ItemsSidebar`.

In the meantime, I have since added a `submitting` state to this component. This should fire immediately. If not, we will have to manually tell React which state update has to happen first.